### ISSUE TITLE
Invitation to Brainstorm for better MQTT BRIDGE feature

### ISSUE THREAD

**Issue by ioolkos** [vernemq/vernemq#1563]
## Context

I'd like to improve the MQTT bridge feature. Both in terms of the implementation but also in discussing best ways for something like MQTT federation between clusters.
The current bridge feature has multiple characteristics that could be improved:
- A flexible but complex client state machine
- A memory only buffer mechanism
- A somewhat unintuitive commandline interface
- Only restricted realtime insights from command line interface
- No dynamic adding of topic patterns.
- ID based on remote IP and Port, so no multiple bridges between 2 nodes without additional listeners.

The bridge is also always a single MQTT client connection. This is fine for forwarding a subset of messages to another cluster, but not an optimal cluster-to-cluster forwarding mechanism for *all* messages or topics.

## Push Bridge
Current specifics: uses internal Subscribe (qos-less, or similar to QoS 0) for local subscriptions, uses MQTT client to publish to remote node.

## Pull Bridge
Current Specifics: Subscribes to remote node with configurable QoS. Can use shared subscriptions from different source nodes.

## Reconnects & Buffering
Current specifics: on receiving a TCP socket close or error, bridge will try to reconnect with a configurable re-connect window. Bridge has a configurable in-memory buffer, ie a queue for messages collected when the bridge is online.

**Comment by ioolkos** [comment]
Related: https://github.com/vernemq/vernemq/pull/1564

**Comment by codeadict** [comment]
About the concurrency, I was thinking if there could be multiple MQTT clients in a process group, one for each topic match:

`vmq_bridge.tcp.br0.topic.1 = demo/+ both 1` - Rus in  process one

`vmq_bridge.tcp.br0.topic.1 = devices/+ both 0` - Runs in process tow.

This will help with sending things out faster in theory and also receiving messages. You can use a general wildcard and still have one process but gives the user flexibility to design for concurrency if it's an issue for them.

The buffer could still be shared among processes somehow.

**Comment by codeadict** [comment]
The RabbitMQ Federation plugin (https://www.rabbitmq.com/federation.html) could be something for inspiration here.

**Comment by ioolkos** [comment]
@codeadict interesting remark on process groups.

I've enabled identifying bridges by {Name, Host, Port} instead of {Host, Port} only, in upcoming PR. This allows for multiple bridges between the same 2 nodes, allowing for 'manual concurrency'.

2 bridges to the same remote listener then look like this, running separately under the bridge supervisor:

![bridges](https://user-images.githubusercontent.com/282359/87314747-f7320880-c523-11ea-939c-dd705b719ba4.png)

**Comment by codeadict** [comment]
Ahh that is also a good approach, thought that may get the user to use 2 bridges that pub/sub to the same topic tree if they are not careful

**Comment by ioolkos** [comment]
Exactly. Also, multiple bridges is only useful when the topics can be partitioned over them. This also has to be ensured manually.

**Comment by codeadict** [comment]
yeah maybe worth in another iteration to partition by topics instead of multiple bridges

**Comment by elduds** [comment]
Bridges are also currently defined on the node not the cluster, so the topic bridging service it provides disappears when a cluster node appears. Some way to make bridge definitions highly available across cluster nodes would be great

**Comment by ioolkos** [comment]
@elduds HA Bridges. I like the sound of that :)

**Comment by codeadict** [comment]
Yep, on the RabbitMQ federation, if a node with a bridge dies, the bridge gets started on another node with the same state. This is gonna need a lot of work to keep bidge state as well as the queued messages replicated.

**Comment by ioolkos** [comment]
@codeadict jep. 
Another approach would be to look at this from the perspective of queues (comparing this to HA queues or Quorum queues in Rabbit rather than federation).
A bridge could be "silently" subscribed to topics on additional cluster node(s) and already have messages. Live bridge would forward status regularly, so that follower bridges can delete messages.
